src/include/cpu/rpp_cpu_simd.hpp

```diff
@@ -10,45 +10,397 @@
+    px3 = _mm_loadu_si128((__m128i *)(srcPtrTemp + 36));    /* load [R13|G13|B13|R14|G14|B14|R15|G15|B15|R16|G16|B16|R17|G17|B17|R18] - Need RGB 13-16 \ */ \
+    px0 = _mm_shuffle_epi8(px0, mask);    /* shuffle to get [R01|R02|R03|R04|G01|G02|G03|G04 || B01|B02|B03|B04|R05|G05|B05|R06] - Need R01-04, G01-04, B01-04 */ \
+    px1 = _mm_shuffle_epi8(px1, mask);    /* shuffle to get [R05|R06|R07|R08|G05|G06|G07|G08 || B05|B06|B07|B08|R09|G09|B09|R10] - Need R05-08, G05-08, B05-08 */ \
+    px2 = _mm_shuffle_epi8(px2, mask);    /* shuffle to get [R09|R10|R11|R12|G09|G10|G11|G12 || B09|B10|B11|B12|R13|G13|B13|R14] - Need R09-12, G09-12, B09-12 */ \
+    px3 = _mm_shuffle_epi8(px3, mask);    /* shuffle to get [R13|R14|R15|R16|G13|G14|G15|G16 || B13|B14|B15|B16|R17|G17|B17|R18] - Need R13-16, G13-16, B13-16 */ \
+    px4 = _mm_unpackhi_epi8(px0, zero);    /* unpack 8 hi-pixels of px0 */ \
+    px5 = _mm_unpackhi_epi8(px1, zero);    /* unpack 8 hi-pixels of px1 */ \
+    px6 = _mm_unpackhi_epi8(px2, zero);    /* unpack 8 hi-pixels of px2 */ \
+    px7 = _mm_unpackhi_epi8(px3, zero);    /* unpack 8 hi-pixels of px3 */ \
+    px0 = _mm_unpacklo_epi8(px0, zero);    /* unpack 8 lo-pixels of px0 */ \
+    px1 = _mm_unpacklo_epi8(px1, zero);    /* unpack 8 lo-pixels of px1 */ \
+    px2 = _mm_unpacklo_epi8(px2, zero);    /* unpack 8 lo-pixels of px2 */ \
+    px3 = _mm_unpacklo_epi8(px3, zero);    /* unpack 8 lo-pixels of px3 */ \
+    p0R = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px0, zero));    /* unpack 4 lo-pixels of px0 - Contains R01-04 */ \
+    p1R = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px1, zero));    /* unpack 4 lo-pixels of px1 - Contains R05-08 */ \
+    p2R = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px2, zero));    /* unpack 4 lo-pixels of px2 - Contains R09-12 */ \
+    p3R = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px3, zero));    /* unpack 4 lo-pixels of px3 - Contains R13-16 */ \
+    p0G = _mm_cvtepi32_ps(_mm_unpackhi_epi16(px0, zero));    /* unpack 4 hi-pixels of px0 - Contains G01-04 */ \
+    p1G = _mm_cvtepi32_ps(_mm_unpackhi_epi16(px1, zero));    /* unpack 4 hi-pixels of px1 - Contains G05-08 */ \
+    p2G = _mm_cvtepi32_ps(_mm_unpackhi_epi16(px2, zero));    /* unpack 4 hi-pixels of px2 - Contains G09-12 */ \
+    p3G = _mm_cvtepi32_ps(_mm_unpackhi_epi16(px3, zero));    /* unpack 4 hi-pixels of px3 - Contains G13-16 */ \
+    p0B = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px4, zero));    /* unpack 4 lo-pixels of px4 - Contains B01-04 */ \
+    p1B = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px5, zero));    /* unpack 4 lo-pixels of px5 - Contains B05-08 */ \
+    p2B = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px6, zero));    /* unpack 4 lo-pixels of px6 - Contains B09-12 */ \
+    p3B = _mm_cvtepi32_ps(_mm_unpacklo_epi16(px7, zero));    /* unpack 4 lo-pixels of px7 - Contains B13-16 */ \
+}
+
+#define RPP_STORE48_F32PLN3_TO_U8PLN3(dstPtrTempR, dstPtrTempG, dstPtrTempB) \
+{ \
+    px4 = _mm_cvtps_epi32(p0R);    /* convert to int32 for R */ \
```

Comment:
what is p0R, p1R..? Better to make #define generic with all arguments passed

Reply:
I have changed it now to take all arguments to completely avoid any ambiguity. (There were 12-16 arguments in a few of them, so I figured if RPP always uses px0/1/2/3 for __m128i and p0/1/2/3 for __m128 as a protocol, we could skip passing those arguments)

Reply:
px0 through px7 is reassigned inside the macro. In that case, please avoid passing them through macro and declare as local variables inside. I think in this case it is better to make it as inline function instead.
You can have a common load macro which would take a "function name" and variable number of arguments, which can call different load routines inside the macro, so the calling macro will be the same.

Reply:
Done. I have added this cleaner handling with variadic inline functions and parameter pack templating now. Added more info on the email.